### ISSUE TITLE
panic when parsing methods on exported fields

### ISSUE THREAD

**Issue by bradleyfalzon** [golang/lint#222]
The following code causes a panic in golint:

```go
package main

import "log"

func (l *log.Logger) Foo() {}
```

The code incorrectly tries to add a method to an external package, the compiler will not accept this `./main.go:5: cannot define new methods on non-local type log.Logger`.

```
panic: interface conversion: ast.Expr is *ast.SelectorExpr, not *ast.Ident

goroutine 1 [running]:
panic(0x6ba820, 0xc820018d80)
        /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/golang/lint.receiverType(0xc820010ea0, 0x0, 0x0)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/lint.go:1444 +0x1ab
github.com/golang/lint.(*pkg).scanSortable.func1(0x7fe5c18841b8, 0xc820010ea0, 0x6cdd01)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/lint.go:336 +0x93
github.com/golang/lint.walker.Visit(0xc8200c08a0, 0x7fe5c18841b8, 0xc820010ea0, 0x0, 0x0)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/lint.go:1474 +0x40
go/ast.Walk(0x7fe5c1884100, 0xc8200c08a0, 0x7fe5c18841b8, 0xc820010ea0)
        /usr/local/go/src/go/ast/walk.go:52 +0x56
go/ast.walkDeclList(0x7fe5c1884100, 0xc8200c08a0, 0xc82000eea0, 0x2, 0x2)
        /usr/local/go/src/go/ast/walk.go:38 +0xd5
go/ast.Walk(0x7fe5c1884100, 0xc8200c08a0, 0x7fe5c1883b70, 0xc82007e580)
        /usr/local/go/src/go/ast/walk.go:353 +0x29b9
github.com/golang/lint.(*file).walk(0xc820018700, 0xc8200c08a0)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/lint.go:1450 +0x80
github.com/golang/lint.(*pkg).scanSortable(0xc820017590)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/lint.go:341 +0x22a
github.com/golang/lint.(*pkg).lint(0xc820017590, 0x0, 0x0, 0x0)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/lint.go:149 +0x75
github.com/golang/lint.(*Linter).LintFiles(0xc8200b37e8, 0xc8200b3958, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/lint.go:110 +0x4c9
main.lintFiles(0xc82000ab10, 0x1, 0x1)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/golint/golint.go:92 +0x2ca
main.lintImportedPackage(0xc82005b500, 0x0, 0x0)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/golint/golint.go:136 +0x6ef
main.lintDir(0x70d838, 0x1)
        /home/bradleyf/scripts/go/src/github.com/golang/lint/golint/golint.go:107 +0x60
main.main()
        /home/bradleyf/scripts/go/src/github.com/golang/lint/golint/golint.go:44 +0x83
```

The panic occurs when golint finds a problem, in the above case because there is no comment for an exported method, but golint doesn't expect a selector in the receiver.

Offending function:

```go
func receiverType(fn *ast.FuncDecl) string {
        switch e := fn.Recv.List[0].Type.(type) {
        case *ast.Ident:
                return e.Name
        case *ast.StarExpr:
                return e.X.(*ast.Ident).Name // panic: interface conversion: ast.Expr is *ast.SelectorExpr, not *ast.Ident
        }
        panic(fmt.Sprintf("unknown method receiver AST node type %T", fn.Recv.List[0].Type))
}
```

I'm not certain on golint's position on fixing this is, but the panic isn't descriptive, and may better be handled by the subsequent panic, such as with the following type assert:

```go
func receiverType(fn *ast.FuncDecl) string {
    switch e := fn.Recv.List[0].Type.(type) {
    case *ast.Ident:
        return e.Name
    case *ast.StarExpr:
        if ident, ok := e.X.(*ast.Ident); ok { // allow subsequent panic to provide a more descriptive error
            return ident.Name
        }
    }
    panic(fmt.Sprintf("unknown method receiver AST node type %T", fn.Recv.List[0].Type)) // panic: unknown method receiver AST node type *ast.StarExpr
}
```

There's obviously other solutions to this by handling a `*ast.SelectorExpr`, such as the following, but it only fixes this specific case, others could exist too which would warrant even more tests.

```go
func receiverType(fn *ast.FuncDecl) string {
        switch e := fn.Recv.List[0].Type.(type) {
        case *ast.Ident:
                return e.Name
        case *ast.StarExpr:
                switch ident := e.X.(type) {
                case *ast.Ident:
                        return ident.Name
                case *ast.SelectorExpr:
                        // ident.X.(*ast.Ident).Name                    == "log"
                        // ident.Sel.Name                               == "Logger"
                        // ident.X.(*ast.Ident).Name+"."+ident.Sel.Name == "log.Logger" - although this does work due to other reasons
                return ident.Sel.Name
                }
        }
        panic(fmt.Sprintf("unknown method receiver AST node type %T", fn.Recv.List[0].Type))
}
```

If this isn't a wont-fix, I'd be happy to PR the preferred fix it if you like.

Note similar discussions have occurred before in regards to whether this is golint's responsibility, but it's a slightly different case, see #134.

**Comment by dmportella** [comment]
I am also seeing this issue.

**Comment by dsnet** [comment]
Fixed with 024c6ac8252128faebc65e3c7dfd295ed7e35674
